### PR TITLE
EIP 101: big-int precompile

### PR DESCRIPTION
**Pull request by pirapira**
If merged, this would solve #235 as part of the Metropolis changes.

Merging this requires the community's consensus on https://github.com/ethereum/EIPs/issues/101

### CODE REVIEW

File: Paper.tex [hunk 1]
```diff
@@ -1393,13 +1398,109 @@ \section{Precompiled Contracts}\label{app:precompiled}
...
+The fourth contract, the identity function $\Xi_{\mathtt{ID}}$ simply defines the output as the input:
 \begin{eqnarray}
 \Xi_{\mathtt{ID}} &\equiv& \Xi_{\mathtt{PRE}} \quad \text{where:} \\
 g_r &=& 15 + 3\Big\lceil \dfrac{|I_\mathbf{d}|}{32} \Big\rceil\\
 \mathbf{o} &=& I_\mathbf{d}
 \end{eqnarray}
 
+The fifth contract performs arbitrary-precision addition on non-negative integers.
+The first word in the input specifies the number of bytes that the first non-negative integer $X$ occupies.
+The result is represented in the smallest possible number of bytes.  The first byte in the output is never zero.
```
Comment (gavofyork) [inline Paper.tex]: ...the first byte in the output, if it exists...

File: Paper.tex [hunk 2]
```diff
@@ -1393,13 +1398,109 @@ \section{Precompiled Contracts}\label{app:precompiled}
...
+
+\begin{eqnarray}
+\Xi_{\mathtt{ADD}} &\equiv& \Xi_{\mathtt{PRE}} \quad \text{except:}\\
+\Xi_{\mathtt{ADD}}(\boldsymbol{\sigma}, g, I) &\equiv& (\varnothing, 0, A^0, ()) \quad \text{if} \quad |I_\mathbf{d}| - 32 <  I_\mathbf{d}[0..31] \\
+g_r &=& G_{addsubbase} + G_{arithword} \Big\lceil \dfrac{|I_\mathbf{d}|}{32} \Big\rceil \\
+\mathbf{o} &=& (X + Y) \in \mathbb{B}_\ell \\
```
Comment (gavofyork) [inline Paper.tex]: use `BE` or something similar to define integer -> bytes.

File: Paper.tex [hunk 3]
```diff
@@ -1393,13 +1398,109 @@ \section{Precompiled Contracts}\label{app:precompiled}
...
 \mathbf{o} &=& I_\mathbf{d}
 \end{eqnarray}
 
+The fifth contract performs arbitrary-precision addition on non-negative integers.
+The first word in the input specifies the number of bytes that the first non-negative integer $X$ occupies.
+The result is represented in the smallest possible number of bytes.  The first byte in the output is never zero.
+
+\begin{eqnarray}
+\Xi_{\mathtt{ADD}} &\equiv& \Xi_{\mathtt{PRE}} \quad \text{except:}\\
+\Xi_{\mathtt{ADD}}(\boldsymbol{\sigma}, g, I) &\equiv& (\varnothing, 0, A^0, ()) \quad \text{if} \quad |I_\mathbf{d}| - 32 <  I_\mathbf{d}[0..31] \\
```
Comment (gavofyork) [inline Paper.tex]: formalise conversion bytes -> integer

File: Paper.tex [hunk 4]
```diff
@@ -1393,13 +1398,109 @@ \section{Precompiled Contracts}\label{app:precompiled}
...
+Y &=& I_\mathbf{d}[(32 + I_\mathbf{d}[0..31])..(|I_\mathbf{d}| - 1)] \\
+\ell_Y &=& |I_\mathbf{d}| - 32 - \ell_X \\
+\ell &=&
+\begin{cases}
+0 &\text{if}\ Y = 0 \ \vee \ \Big\lfloor \dfrac X Y \Big\rfloor = 0 \\
+\lfloor \log_8(\Big\lfloor \dfrac X Y \Big\rfloor) \rfloor + 1 & \text{otherwise}
+\end{cases}
+\end{eqnarray}
+
+The ninth contract performs arbitrary-precision exponentiation under modulo.  Here, $0 ^ 0$ is taken to be one.
```
Comment (gavofyork) [inline Paper.tex]: mention how the arguments are encoded here.

File: Paper.tex (hunk 2 above)
Comment (pirapira) [inline Paper.tex]: * the symbol `\ell` is already taken
* the number of bits should be `8 \ell` not `\ell`.

**Comment by pirapira** [comment]
To be reopend on `metropolis` branch.

File: Paper.tex (hunk 3 above)
Comment (pirapira) [inline Paper.tex]: * Maybe `I_\mathbf{d}[0..31]` can be assigned to a variable.
* Or the common parsing logic should be factored out.

File: Paper.tex [hunk 5]
```diff
@@ -1393,13 +1398,109 @@ \section{Precompiled Contracts}\label{app:precompiled}
...
+0 &\text{if}\, X = Y = 0 \\
+\lfloor \log_8(X + Y) \rfloor + 1 & \text{otherwise}
+\end{cases}
+\end{eqnarray}
+
+The sixth contract performs arbitrary-precision subtraction on non-negative integers.  This is similar to the addition.  The subtraction contract halts exceptionally if the result would be negative.
+
+\begin{eqnarray}
+\Xi_{\mathtt{SUB}} &\equiv& \Xi_{\mathtt{PRE}} \quad \text{except:}\\
+\Xi_{\mathtt{SUB}}(\boldsymbol{\sigma}, g, I) &\equiv& (\varnothing, 0, A^0, ()) \quad \text{if} \quad |I_\mathbf{d}| - 32 <  I_\mathbf{d}[0..31] \\
```
Comment (pirapira) [inline Paper.tex]: * `... if ...; ... if ...` should go into `begin{cases} \end{cases}`

**Pull request closed without merging**